python/pyarrow/parquet.py

```diff
@@ -591,23 +592,32 @@ class ParquetManifest(object):
 
     """
     def __init__(self, dirpath, filesystem=None, pathsep='/',
-                 partition_scheme='hive'):
+                 partition_scheme='hive', thread_pool=None):
         self.filesystem = filesystem or _get_fs_from_path(dirpath)
         self.pathsep = pathsep
         self.dirpath = dirpath
         self.partition_scheme = partition_scheme
         self.partitions = ParquetPartitions()
         self.pieces = []
+        self._thread_pool = thread_pool or futures.ThreadPoolExecutor()
 
         self.common_metadata_path = None
         self.metadata_path = None
 
         self._visit_level(0, self.dirpath, [])
 
+        # Due to concurrency, pieces will potentially by out of order if the
+        # dataset is partitioned so we sort them to yield stable results
+        self.pieces.sort(key=lambda piece: piece.path)
+
         if self.common_metadata_path is None:
             # _common_metadata is a subset of _metadata
             self.common_metadata_path = self.metadata_path
 
+        if thread_pool is None:
+            # If we created the thread pool make sure to shut it down after use
+            self._thread_pool.shutdown()
```

Comment:
I may be missing something, but above `thread_pool` defaults to `None` and I see where `_thread_pool` is set on line 602. What is being `shutdown()` here if `thread_pool` is `None`?

Reply:
If `thread_pool is None` that means we instantiated it on line 602 and therefore the calling user will not be using that same thread pool so we should properly shut it down to preserve resources. If it wasn't none that means the user passed in a thread pool executor likely to reuse an existing one that already existed.

Reply:
This ensures that a user-supplied thread pool is not shut down; if one was created above then it should be closed now